opal/mca/btl/ofi/btl_ofi_component.c

```diff
@@ -0,0 +1,583 @@
+
+    /* Select the provider */
+    fabric_attr.prov_name = prov_include;
+
+    domain_attr.mr_mode = MCA_BTL_OFI_REQUESTED_MR_MODE;
+
+    /* message progression mode. */
+    if (!strcmp(ofi_progress_mode, "auto")) {
+        progress_mode = FI_PROGRESS_AUTO;
+    } else if (!strcmp(ofi_progress_mode, "manual")) {
+        progress_mode = FI_PROGRESS_MANUAL;
+    } else {
+        progress_mode = FI_PROGRESS_UNSPEC;
+    }
+
+    domain_attr.control_progress = progress_mode;
+    domain_attr.data_progress = progress_mode;
+
+    /* select endpoint type */
+    ep_attr.type = FI_EP_RDM;
+
+    /* ask for capabilities */
+    hints.caps = MCA_BTL_OFI_REQUIRED_CAPS;
+
+    hints.fabric_attr = &fabric_attr;
+    hints.domain_attr = &domain_attr;
+    hints.ep_attr = &ep_attr;
+    hints.tx_attr = &tx_attr;
+    hints.rx_attr = &rx_attr;
+
+    /* for now */
+    tx_attr.iov_limit = 1;
+    rx_attr.iov_limit = 1;
+
+    mca_btl_ofi_component.module_count = 0;
+
+    /* do the query. */
+    rc = fi_getinfo(libfabric_api, NULL, NULL, 0, &hints, &info_list);
+    if (0 != rc) {
+        BTL_VERBOSE(("fi_getinfo failed with code %d: %s",rc, fi_strerror(-rc)));
+        return NULL;
+    }
+
+    /* count the number of resources/ */
+    info = info_list;
+    while(info) {
+        resource_count++;
+        info = info->next;
+    }
+    BTL_VERBOSE(("ofi btl found %d possible resources.", resource_count));
+
+    info = info_list;
+
+    while(info) {
+        rc = validate_info(info);
+        if (OPAL_SUCCESS == rc) {
+            /* Device passed sanity check, let's make a module.
+             * We only pick the first device we found valid */
+            rc = mca_btl_ofi_init_device(info);
+            if (OPAL_SUCCESS == rc)
+                break;
+        }
+        info = info->next;
+    }
+
+    /* We are done with the returned info. */
+    fi_freeinfo(info_list);
+
+    /* pass module array back to caller */
+    base_modules = calloc (mca_btl_ofi_component.module_count, sizeof (*base_modules));
+    if (NULL == base_modules) {
+        return NULL;
+    }
+
+    memcpy(base_modules, mca_btl_ofi_component.modules,
+           mca_btl_ofi_component.module_count *sizeof (mca_btl_ofi_component.modules[0]));
+
+    BTL_VERBOSE(("ofi btl initialization complete. found %d suitable transports",
+                 mca_btl_ofi_component.module_count));
+
+    *num_btl_modules = mca_btl_ofi_component.module_count;
+
+    return base_modules;
+}
+
+static int mca_btl_ofi_init_device(struct fi_info *info)
+{
+    int rc;
+    int *module_count = &mca_btl_ofi_component.module_count;
+    size_t namelen;
+    mca_btl_ofi_module_t *module;
+
+    char *linux_device_name;
+    char ep_name[FI_NAME_MAX];
+    struct fi_info *ofi_info;
+    struct fi_cq_attr cq_attr = {0};
+    struct fi_av_attr av_attr = {0};
+    struct fid_fabric *fabric;
+    struct fid_domain *domain;
+    struct fid_ep *endpoint;
+    struct fid_cq *cq;
+    struct fid_av *av;
```

Comment:
I think you need to initialize all of these pointers to NULL to make sure your safety conditionals in the `fail` block work as expected.  😉

Reply:
You kids and your C99... (I'm not actually 100% sure if stack variables are guaranteed to be 0'ed out or not...?).

Reply:
I know, we have it easy. 
No but really, does Open MPI build without C99?

Reply:
No, OMPI's configure enforces building with a C99-compliant compiler.

Reply:
stack variables are not zero by default